### ISSUE TITLE
[pilot 1.16.0] envoyExtAuthzHttp doesn't forward `x-envoy-peer-metadata` headers

### ISSUE THREAD

**Issue by parkedwards** [istio/istio#42210]
### Bug Description

TL;DR:

1. `envoyExtAuthzHttp` subrequests are consistently showing `source_app=unknown`
2. to remedy this, we started forwarding `x-envoy-peer-metadata` to the authorization service
3. after 1.16.0, we do not see these `x-envoy-peer-metadata` forwarded any longer, when enabling debug envoy logs

Hello there,

We've been running `pilot` on `1.15.3`, and all of our applications have the `istio-proxy` envoy sidecar injected ✅ 

We use prometheus / kiali to monitor, and for the most part - a metric like `istio_requests_total` correctly shows `source_app` and `destination_app`.  ✅ 

HOWEVER, there is a subset of cases where `source_app=unknown, reporter=destination` ❌ :
* we currently use `envoyExtAuthzHttp` to "externally" authorize each application request (the authorizer is an in-cluster auth service)
* all applications + auth server have the istio-proxy -- calls from ingress -> auth directly are correctly tagged in telemetry
* the ext_authz subreqests are showing "unknown" `source_app` values
* in the subrequests specifically, we're seeing something like:

```
# example prometheus counter for the ext_authz request, specifically

istio_requests_total{destination_app="auth", source_app="unknown", source_principal="<service account of destination app>"}
```

To "remedy" these cases (for our observability + sanity), we configured our [envoyExtAuthzHttp](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig-ExtensionProvider-EnvoyExternalAuthorizationHttpProvider) meshConfig settings to forward the peer-metadata headers -- the intention was to tag these ext_authz subrequests to be from the original, upstream requester (normally our ingress gateway)

```yaml
        - name: "authz-http-redirect"
          envoyExtAuthzHttp:
            port: "<auth service port>"
            service: "auth.default.svc.cluster.local"
            includeRequestHeadersInCheck:
              - "x-envoy-peer-metadata"
              - "x-envoy-peer-metadata-id"
```

However, since updating pilot to 1.16.0, we're noticing (via debug logs on the destination app + auth) that the `x-envoy-peer-metadata.*` headers are no longer being forwarded via `envoyExtAuthzHttp`, which breaks this "fix" around an unknown request hitting our authorization service

Is that expected - were we using these headers in an incorrect way pre 1.16?  If so, was there a correct way to remedy this monitoring issue when using `envoyExtAuthzHttp`?

### Version

```prose
(not using `istioctl`, we are using the `istiod` helm chart for discovery)

discovery: 1.16.0 (distroless)
sidecar proxies: 1.16.0 (distroless)
ingress gateway: 1.16.0 (distroless)

$ kubectl version --short
Client Version: v1.25.1
Kustomize Version: v4.5.7
Server Version: v1.24.6-gke.1500
```

**Comment by parkedwards** [comment]
also for record-keeping -- I opened a similar thread in the istio slack workspace here:

https://istio.slack.com/archives/C382V8Q92/p1669665387728149

**Comment by zirain** [comment]
can you share the config_dump?

**Comment by parkedwards** [comment]
@zirain i've attached a config_dump for the destination pod envoy (which I believe kicks off the ext_authz call)

the request flow would be something like:

ingress gateway -> destination envoy -> (ext_authz) auth service -> destination

if some of the values look strange, it's bc I attempted to redact any sensitive info with nonsense.  but the `envoy.filters.http.ext_auth` configurations should be un-edited -- you can see where we specify the `peer-metadata` header forwarding

[config_dump.json.zip](https://github.com/istio/istio/files/10128274/config_dump.json.zip)

**Comment by kyessenov** [comment]
I think the order was changed to put MX filter first so that rbac could see peer metadata. That same filter also strips the header, so that's a side-effect. The easy workaround is to put extauthz first. Come to think of it, I think Tetrate's distribution might be affected @lizan .

**Comment by zirain** [comment]
@kyessenov related to https://github.com/istio/istio/pull/41261?

**Comment by kyessenov** [comment]
Correct, @zirain .

**Comment by parkedwards** [comment]
> The easy workaround is to put extauthz first

@kyessenov was this directed at us operators?  if so, our `ext_authz` is configured/enabled via `meshConfig.extensionProviders` -- are you saying we should be placing it first in this list?

**Comment by kyessenov** [comment]
@parkedwards This is a regression from the operator point of view. 
My comment was addressed to @hzxuzhonghu @howardjohn , the PR needs to be reverted or be guarded by a flag.

**Comment by howardjohn** [comment]
I think before or after breaks different people depending on how they want things ordered/read the MX (header or filter state). I don't think there is a universal correct approach. Given the header is an implementation detail relying on it seems suspicious though.

**Comment by parkedwards** [comment]
> Given the header is an implementation detail relying on it seems suspicious though.

I'd agree with that.  I'd prefer not having to forward the peer-metadata header to `ext_authz` calls if possible

but is there a way to preserve telemetry `source_app` value(s) on those `ext_authz` calls?  without explicitly propagating the metadata header, these subrequests all show as `source_app=unknown` for us

**Comment by kyessenov** [comment]
The correct way is to forward the internal request state (`wasm.downstream_peer`) but the Istio API doesn't have it. MX filter communicates with other filters via this state.

**Comment by parkedwards** [comment]
interesting. so anyone using this `ext_authz` filter via istio will experience the `source_app=unknown` telemetry?

**Comment by parkedwards** [comment]
> The correct way is to forward the internal request state (wasm.downstream_peer) but the Istio API doesn't have it. MX filter communicates with other filters via this state.

@kyessenov so based on your comment - is there no way to propagate peer metadata when using the external authorization filter?

**Comment by parkedwards** [comment]
@lei-tang got it

so am i correct in understanding that after 1.16, no external authorization calls will have a known `source_app` or `destination_app` value (depending on the reporter)?

**Comment by kyessenov** [comment]
@costinm Another instance of header operation re-ordering ^^^

**Comment by BenCoughlan15** [comment]
Hey, is there any workaround or fix for this? 

We were using upstream headers from authz, as part of the virtual service routing decision, to define auth and regional routing. 

This change breaks auth for us entirely, as @howardjohn mentioned, it breaks functionality for us this way, and for other people the other way, so shouldn't this be configurable? 

B

**Comment by chr15murray** [comment]
I work with @BenCoughlan15 and yesterday we explored duplicating the istio.metadata_exchange filter by adding the following EnvoyFilter

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  labels:
    istio.io/rev: canary
  name: metadata-exchange-after-authz
  namespace: istio-system
spec:
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: GATEWAY
      listener:
        filterChain:
          filter:
            name: envoy.filters.network.http_connection_manager
            subFilter:
              name: envoy.filters.http.ext_authz
    patch:
      operation: INSERT_AFTER
      value:
        name: istio.metadata_exchange
        typed_config:
          '@type': type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
          config:
            configuration:
              '@type': type.googleapis.com/envoy.tcp.metadataexchange.config.MetadataExchange
            vm_config:
              code:
                local:
                  inline_string: envoy.wasm.metadata_exchange
              runtime: envoy.wasm.runtime.null
  workloadSelector:
    labels:
      istio: ingressgateway
```

This results in the following filter chain order:
```
"istio.metadata_exchange"
"envoy.filters.http.rbac"
"envoy.filters.http.ext_authz"
"istio.metadata_exchange"
"istio.alpn"
"envoy.filters.http.fault"
"envoy.filters.http.cors"
"istio.stats"
"envoy.filters.http.router"
```

At first glance this appears to be a workaround and we are able to use headers from http.ext_authz that are part of the `headersToUpstreamOnAllow` to make routing decisions in a virtual service:
```
  - directResponse:
      body:
        string: "header from ext_authz"
      status: 200
    match:
    - headers:
        aauthztest:
          exact: test
```

@howardjohn - are you aware of any downsides of calling this twice? 

We are also exploring if the use of envoys dynamic metadata might also allow us to do this as this might be a cleaner solution.
